client/rpc/src/debug.rs

```diff
@@ -0,0 +1,207 @@
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+// This file is part of Frontier.
+//
+// Copyright (c) 2023 Parity Technologies (UK) Ltd.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use std::{marker::PhantomData, sync::Arc};
+
+use ethereum::EnvelopedEncodable;
+use ethereum_types::H256;
+use jsonrpsee::core::{async_trait, RpcResult};
+use rlp::Encodable;
+// Substrate
+use sc_client_api::backend::{Backend, StorageProvider};
+use sp_api::ProvideRuntimeApi;
+use sp_blockchain::HeaderBackend;
+use sp_runtime::traits::Block as BlockT;
+// Frontier
+use fc_rpc_core::{types::*, DebugApiServer};
+use fc_storage::OverrideHandle;
+use fp_rpc::EthereumRuntimeRPCApi;
+
+use crate::{cache::EthBlockDataCacheTask, frontier_backend_client, internal_err};
+
+/// Debug API implementation.
+pub struct Debug<B: BlockT, C, BE> {
+	client: Arc<C>,
+	backend: Arc<dyn fc_api::Backend<B>>,
+	overrides: Arc<OverrideHandle<B>>,
+	block_data_cache: Arc<EthBlockDataCacheTask<B>>,
+	_marker: PhantomData<BE>,
+}
+
+impl<B: BlockT, C, BE> Debug<B, C, BE> {
+	pub fn new(
+		client: Arc<C>,
+		backend: Arc<dyn fc_api::Backend<B>>,
+		overrides: Arc<OverrideHandle<B>>,
+		block_data_cache: Arc<EthBlockDataCacheTask<B>>,
+	) -> Self {
+		Self {
+			client,
+			backend,
+			overrides,
+			block_data_cache,
+			_marker: PhantomData,
+		}
+	}
+
+	async fn block_by(&self, number: BlockNumberOrHash) -> RpcResult<Option<ethereum::BlockV2>>
```

Comment:
Methods like `block_by`, `transaction_by`, `receipts_by` have a lot of duplicate code with the methods in the eth module. I plan to add a helpful mapping backend type in the future RPs to reduce duplicate code and improve maintainability.